File: src/main/java/juuxel/cubic/creature/Creature.java
/* This file is a part of the Cubic source code,
 * created by Juuxel.
 *
 * Cubic is licensed under the GNU LGPLv3 license.
 * Full source and license: https://github.com/Juuxel/Cubic
 */
package juuxel.cubic.creature;

import juuxel.cubic.Cubic;
import juuxel.cubic.ecs.AbstractEntity;
import juuxel.cubic.ecs.components.Components;
import juuxel.cubic.event.CreatureCreationEvent;
import juuxel.cubic.event.EventBus;
import juuxel.cubic.lib.GameValues;
import juuxel.cubic.render.Graphics;
import juuxel.cubic.render.sprite.Sprite;
import juuxel.cubic.util.Direction;
import juuxel.cubic.util.Utils;

import java.awt.image.BufferedImage;
import java.util.Random;

public abstract class Creature extends AbstractEntity
{
    @Deprecated // PositionComponent
    protected double x, y;

    @Deprecated // SpeedComponent
    public double xSpeed = 0, ySpeed = 0;

    protected final Random random = new Random();

    @Deprecated // CollisionBoxComponent, SpriteComponent
    // Bad names; they are used for more than sprites
    public int spriteWidth = 32, spriteHeight = 32;

    @Deprecated // SpriteComponent
    private Sprite sprite;

    @Deprecated // GroundCollisionComponent, DirectionComponent
    public boolean collidesWithGround, flippingEnabled;

    @Deprecated // DirectionComponent
    public Direction direction = Direction.RIGHT;

    @Deprecated // I think this can be removed (initially SpeedComponent)
    protected int speedModifierX = 1, speedModifierY = 1;

    protected Creature()
    {
        Cubic.CREATURES.add(this);
        EventBus.post(new CreatureCreationEvent(this));
        collidesWithGround = false;
        flippingEnabled = false;
    }

    @Override
    protected void attachComponents()
    {
        attachComponent(Components.POSITION);
        attachComponent(Components.SPEED);
        // TODO Add only to creatures that collide
        attachComponent(Components.GROUND_COLLISION);
        // TODO Add only to creatures that can die
        attachComponent(Components.ALIVE);
    }

    protected double yOnScreen()
    { return Utils.yOnScreen(y); }

    public void kill()
    {
        onComponent(Components.ALIVE, comp -> comp.active = false);
    }

    private void drawCreature(Graphics g, BufferedImage image, int width, int height)
    {
        int dx = (int) x, dy = (int) yOnScreen();

        if (direction == Direction.RIGHT)
            g.drawImage(image, dx - width / 2, dy - height / 2, width, height);
        else
            g.drawFlippedImage(image, dx - width / 2, dy - height / 2, width, height);
    }

    protected void drawCreature(Graphics g, Sprite sprite)
    {
        drawCreature(g, sprite.getImage(this), spriteWidth, spriteHeight);
    }

    @Deprecated
    // "logic"??
    public final void executeLogic()
    {
        moveX(xSpeed / speedModifierX);

        if (collidesWithGround)
        {
            double yNew = y + ySpeed / speedModifierY;

            if (yNew < GameValues.GROUND)
                yNew = GameValues.GROUND;

            y = yNew;

            if (onGround())
                ySpeed = 0;
        }
        else
            y += ySpeed / speedModifierY;

        logic();
    }

    @Deprecated
    // Also, why is this called "logic"?
    protected abstract void logic();

    public void draw(Graphics g)
    {
        if (getSprite() != null)
            drawCreature(g, getSprite());
    }

    public Sprite getSprite()
    { return sprite; }

    protected void setSprite(Sprite sprite)
    {
        this.sprite = sprite;
        this.flippingEnabled = sprite.isFlippable();
    }

    public boolean onGround()
    { return y <= GameValues.GROUND; }

    public void updateSpriteDirection(double value)
    {
        if (flippingEnabled)
        {
            if (value > 0)
                direction = Direction.RIGHT;
            else if (value < 0)
                direction = Direction.LEFT;
        }
    }

    public void updateSpriteDirection()
    {
        updateSpriteDirection(xSpeed);
    }

    public void moveX(double xOffset)
    {
        updateSpriteDirection(xOffset);

        x += xOffset;
    }

    public void moveY(double yOffset)
    { y += yOffset; }

    public double getX()
    { return x; }

    public double getY()
    { return y; }
}

File: src/main/java/juuxel/cubic/creature/Creature.kt
package juuxel.cubic.creature

import juuxel.cubic.Cubic
import juuxel.cubic.lib.GameValues
import juuxel.cubic.lib.Images
import juuxel.cubic.render.Graphics
import juuxel.cubic.render.sprite.Sprite
import juuxel.cubic.util.Direction
import juuxel.cubic.util.IBasicFunctions
import java.awt.Image
import java.util.*

abstract class Creature : IBasicFunctions
{
    public val id: Int = Cubic.CREATURES.size
    public var x: Double = 0.toDouble()
        protected set
    public var y: Double = 0.toDouble()
        protected set
    public var xSpeed = 0.0
    public var ySpeed = 0.0
    public var spriteWidth = 32
    public var spriteHeight = 32
    public var sprite: Sprite = Images.bricks // Just a placeholder, should probably use some proper missing texture thing
    public var isCollisionEnabled: Boolean = false
    public var isFlippingEnabled: Boolean = false
    public var direction = Direction.RIGHT

    protected val random = Random()
    protected var speedModifierX = 1
    protected var speedModifierY = 1

    init
    {
        Cubic.CREATURES.add(this)
        Cubic.CREATURE_LISTENERS.forEach { it.onCreatureCreated(this) }
    }

    fun calculateY(): Double
    {
        return calculateY(y)
    }

    open fun kill()
    {
    }

    private fun drawCreature(g: Graphics, image: Image, width: Int, height: Int)
    {
        val dx = x.toInt()
        val dy = calculateY().toInt()

        if (direction == Direction.RIGHT)
            g.drawImage(image, dx - width / 2, dy - height / 2, width, height)
        else
            g.drawFlippedImage(image, dx - width / 2, dy - height / 2, width, height)
    }

    protected fun drawCreature(g: Graphics, sprite: Sprite)
    {
        drawCreature(g, sprite.getImage(this), spriteWidth, spriteHeight)
    }

    fun executeLogic()
    {
        moveX(xSpeed / speedModifierX)

        if (isCollisionEnabled)
        {
            var yNew = y + ySpeed / speedModifierY

            if (yNew < GameValues.GROUND)
                yNew = GameValues.GROUND

            y = yNew

            if (onGround())
                ySpeed = 0.0
        }
        else
            y += ySpeed / speedModifierY

        logic()
    }

    protected abstract fun logic()

    open fun draw(g: Graphics)
    {
        drawCreature(g, sprite)
    }

    fun onGround(): Boolean
    {
        return y <= GameValues.GROUND
    }

    fun moveX(xOffset: Double)
    {
        if (isFlippingEnabled)
        {
            direction = when {
                xOffset > 0 -> Direction.RIGHT
                xOffset < 0 -> Direction.LEFT
                else -> direction
            }
        }

        x += xOffset
    }

    fun moveY(yOffset: Double)
    {
        y += yOffset
    }
}
